### PR TITLE
SDS Secret Provisioning for Egress Gateway using Destination Rule API

### PR DESCRIPTION
**Pull request by nschhina**
Design Doc: https://docs.google.com/document/d/1--c7Y4MR5CL-q6Fl5UlYLzluQpAlRcdyJ1gOUB98Njg/edit
Ref: https://github.com/istio/istio/issues/14039

### CODE REVIEW

**Comment by nschhina** [comment]
@googlebot I fixed it.

**Review by rshriram — COMMENTED** [review]
the github diff in this PR is very hard to understand given the scope of changes.. Do you want to do a quick PR that moves all the required code (as you seem to be doing) [no changes to tests whatsoever] and then another PR that actually enables the sds in destination rule only for egress gateway [assuming thats what you are doing. If you are simply enabling SDS in destination rule for all, then let us know]

**Comment by nschhina** [comment]
@rshriram I don't see a neat way to split this. Well right now all this PR is doing is generating SDS config from CredentialName(if specified), permission to read is only given to gateway proxy which is configured in sdsservice module(todo). So this one builds tls context with sds for credentialName for `any` proxy type i.e workloads and gw both. I've also moved all TLS context building logic to it's own function and extended testing to rigorously unit test that function. Can you also please override this CLA check, it's complaining as my github email wasn't configured on previous commits

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (JimmyCYJ) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:1046]: Could you add a comment here that this is the networking.ClientTLSSettings_MUTUAL case?

File: pilot/pkg/security/model/authentication.go
Comment (JimmyCYJ) [inline pilot/pkg/security/model/authentication.go:210]: This function needs to check tlsOpts.Mode, just like ApplyCustomSDSToServerCommonTLSContext().
If it is networking.ClientTLSSettings_SIMPLE, then only need to configure tlsContext.ValidationContextType. SIMPLE means TLS handshake, the client side only needs a server side root cert, and use that root cert to validate the server cert sent from the remote server.
The tlsContext.TlsCertificateSdsSecretConfigs() is to get the client side key/cert via SDS, which are sent to the remote server side during a mTLS handshake.

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (nschhina) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml]: yes, I'm using that, I was testing a bug and mounting mechanism so I changed it to see how egress sds generation happens in that case. I've already pushed a new change which reverts it back

File: pilot/pkg/security/model/authentication.go
Comment (nschhina) [reply pilot/pkg/security/model/authentication.go]: Thanks totally missed it! Can you review if it looks good now

**Comment by nschhina** [comment]
cc: @myidpt

File: pkg/istio-agent/sds-agent.go
Comment (howardjohn) [inline pkg/istio-agent/sds-agent.go:66]: Why do we need both ingress and egress? A single istio-agent is either ingress or egress? We should aim to keep config as minimal as possible

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: Actually... now that I think about it it *can* be both ingress and egress. As a result we should actually likely have everything be a single config - same uds, etc. So we don't need to duplicate everything

File: security/pkg/nodeagent/sds/sds_stream_test.go
Comment (howardjohn) [inline security/pkg/nodeagent/sds/sds_stream_test.go:351]: same comment above - this should be merged

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:1010]: this function is massive and hard to follow (before and after your PR). Its confusing to have the switch cover two case (simple and mutual) yet then inside the switch branch on these two.

Can we break this up a bit? Rather than major nesting, we can exit early. For example, if tls.credentialName is set, exit early instead of going into another `else{}` branch

**Review by howardjohn — CHANGES_REQUESTED** [review]
I would expect some integration tests associated with this change

File: pkg/istio-agent/sds-agent.go
Comment (nschhina) [reply pkg/istio-agent/sds-agent.go:66]: I think Security team prefers different configs(uds path etc.) for now, but I'll open an issue and follow up on that issue to merge both in a future PR. cc: @JimmyCYJ

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: Why would we duplicate all of the code? its literally identical.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (nschhina) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Did some code duplication for CredentialName but it should improve readability now

File: pkg/istio-agent/sds-agent.go
Comment (JimmyCYJ) [reply pkg/istio-agent/sds-agent.go:66]: Using a single uds sounds good to me. I would prefer to be conservative to begin with a different uds, to make sure we don't break the ingress sds logic in sds-agent. Later we can have a refactoring PR and combine two uds into one.

Comment (nschhina) [reply pkg/istio-agent/sds-agent.go:66]: So currently, the way we know if it's a ingress gateway proxy is by the existence of /var/run/ingress_gateway directory on the workload. Similarly for egress gw proxy... /var/run/egress_gateway. And most duplication ties back to the sdsudspaths, which I think to avoid any chance of regression we should keep separate for now and defer merging the two in a future PR. At least that's what I understood from the discussion I had with security team.

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: Its not just "refactoring" - its a matter of correctness. There is, in reality, no such thing as "egress" and "ingress" gateways. There are just gateways. What if I want to configure a DR with TLS on my "ingress" gateway? I cannot do it because we decided we wanted to duplicate a bunch of code that is identical?

I also don't see why it would break the ingress logic - it is unchanged entirely. We just activate the codepath by default on egress gw now. Ingress remains unchanged

Comment (nschhina) [reply pkg/istio-agent/sds-agent.go:66]: Ingress and Egress gateway deployment are still two different resources even though they both tie back to the same Gateway API. I don't quite understand what you mean by DR with TLS on ingress gateway. Can you explain what case you're concerned about

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: Lets not confuse the deployments we offer users as part of the default installation with the behavior of the system/APIs. A user can install arbitrary amounts of gateways, which are controlled by the Gateway API (among others). There is absolutely no distinction between "ingress" and "egress" gateways in Pilot code, because there is none in the API or in reality. 

There is client TLS (DR) and server TLS (Gateway) - there is not "egressgateway" tls and "ingressgateway" tls. We can either, or both, on any gateway.

Just because we have a default example of a split ingress and egress gateway does not mean that is how gateways actually work. They are merely examples/best practices to get people started.

Comment (nschhina) [reply pkg/istio-agent/sds-agent.go:66]: If I understand your case correctly, I think the case you are mentioning is ingress gw originating tls, well if that's the case- this feature isn't meant to support that use case currently.
external workload -> ingress gateway -> external workload is not a use case I think we ever see? 
For external workload -> ingress gateway -> sidecar, we aren't even supporting DR to use credentialName to fetch secrets at ingress gateway.

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: > this feature isn't meant to support that use case currently.

Then I disagree with this feature :slightly_smiling_face: . We cannot just merge stuff that applies to small subset of cases - its a horrible user experience to try to debug your way through all of the partially implemented APIs.

You're examples are too limited - applying strict assumptions to a project with as wide a user base as Istio will not hold up in practice. Its not `external workload -> ingress gateway -> external workload`. The flow is `arbitrary workload -> arbitrary gateway -> arbitrary workload`

Comment (nschhina) [reply pkg/istio-agent/sds-agent.go:66]: Ah I agree with you now, my examples use case was too restrictive. `arbitrary workload -> arbitrary gateway -> arbitrary workload` would work best if we use single uds path.

Comment (JimmyCYJ) [reply pkg/istio-agent/sds-agent.go:66]: That's good point. I think what we need is to rename the uds path to be a general one, and let sds-server listens on this uds path, and update the logic in ingressSdsExists() when creating gatewaySecretCache.

Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go:66]: The actual name is less important imo. I don't think a user will complain if some internal name is called `ingress_uds` and its on their egress gateway. Changing the name has real costs (its not currently dynamically configurable) so probably not worth it.

**Review by howardjohn — APPROVED** [review]
This is great! thanks!

**Review by JimmyCYJ — APPROVED** [review]

**Review by JimmyCYJ — APPROVED** [review]

**Comment by nschhina** [comment]
I've signed the CLA.

**Comment by nschhina** [comment]
@googlebot I've signed the CLA.

**Pull request merged**
